src/js/controller/next-item-promise.js

```diff
@@ -0,0 +1,60 @@
+
+export class ItemPromise {
+
+    constructor (index, model, api) {
+        this.index = index;
+        this.model = model;
+        this.api = api;
+        this.promise = new Promise((resolve, reject) => {
+            this.resolve = resolve;
+            this.reject = reject;
+        });
+        this.async = null;
+    }
+
+    set callback (handler) {
+        this.async = handler;
+    }
+
+    run () {
+        const { api, async, index, resolve, reject, promise } = this;
+        const playlistItem = this.getItem(index);
+        if (!playlistItem) {
+            const message = index === -1 ? 'No recs item' : `No playlist item at index ${index}`;
+            reject(new Error(message));
+        }
+        if (async) {
+            this.clear();
+            const asyncPromise = async.call(api, playlistItem, index) || Promise.resolve();
+            asyncPromise.then(resolve).catch(reject);
+        } else {
+            resolve(playlistItem);
+        }
+        return promise;
+    }
+
+    getItem(index) {
+        const { model } = this;
+        if (index === -1) {
+            return model.get('nextUp');
+        }
+        const playlist = model.get('playlist');
+        return playlist[index];
+    }
+
+    clear () {
+        this.async = null;
+    }
+
+    preload () {
+        return this.run();
+    }
+
+    setItem (index) {
+        if (index !== this.index) {
+            this.reject(new Error(`Item ${this.index} was skipped.`));
```

Comment:
this is probably a result of me not fully groking the execution pathing but is `setItem` in this promise controller another method to change the item vs an explicit call to `setPlaylistItem` on the global api?

Reply:
No. `ItemPromise.setItem(index)` is called to run the itemPromise as the program-controller attempts to change the item index. The index check was added to reject skipped items, which looks like it needs more work and testing (we'd need to replace the promise for the next time it's used...)

The idea is that there are two places where run the async callback, when attempting to preload or load the item. So `ItemPromise` has two methods `preload` and `setItem` and they just call run. If we want to extend the arguments we pass to the callback to include which event context we're in we could pass that into `run` from these methods.
```js
    preload () {
        return this.run();
    }

    setItem (index) {

        return this.run();
    }
```